### PR TITLE
[Exp] Cherry-pick warp MDP migration and capture safety from dev/newton

### PR DESCRIPTION
**Pull request by hujc7**
## Summary

* Cherry-picks [Newton] Migrate more envs and mdps to warp (https://github.com/isaac-sim/IsaacLab/pull/4690) onto develop
* Cherry-picks [Newton] Add capture safety guards and fix WrenchComposer stale COM pose (https://github.com/isaac-sim/IsaacLab/pull/4779) onto develop

### Changes included
- Warp-first MDP terms (observations, rewards, events, terminations, actions) for manager-based envs
- Tested warp env configs: Ant, Humanoid, Cartpole, locomotion velocity (A1, AnymalB/C/D, Cassie, G1, Go1/2, H1), Franka/UR10 reach
- ManagerCallSwitch max_mode cap and scene capture config
- MDP kernels made graph-capturable with consolidated test infrastructure
- capture_unsafe safety guards on lazy-evaluated derived properties in articulation/rigid_object data
- WrenchComposer fix: use fresh COM pose buffers instead of stale cached link poses

### Dropped
- G1-29-DOF warp env (Isaac-Velocity-Flat-G1-Warp-v1): removed because the stable g1_29_dofs task config does not exist on develop (only on dev/newton). Warp env PRs should only add warp frontends for envs that already exist in the stable package.

## Dependencies

Must be merged **after** these PRs (in order):
1. https://github.com/isaac-sim/IsaacLab/pull/4905 (merged)
2. https://github.com/isaac-sim/IsaacLab/pull/4829

## Validated base

Validated against develop at 7588fa9ed5f.

## Test plan

- [x] Run warp env training sweep across all manager-based env configs (14/14 pass, mode=2, 4096 envs, 300 iters)
- [ ] Run test_mdp_warp_parity.py and test_mdp_warp_parity_new_terms.py
- [ ] Run test_action_warp_parity.py
- [ ] Verify WrenchComposer COM pose is fresh (not stale) during graph replay

### CODE REVIEW

**Comment by hujc7** [comment]
@greptileai Review

File: source/isaaclab/isaaclab/utils/warp/utils.py [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+        if isinstance(mask, torch.Tensor):
+            if mask.dtype != torch.bool:
+                mask = mask.to(dtype=torch.bool)
+            if str(mask.device) != device:
+                mask = mask.to(device)
+            return wp.from_torch(mask, dtype=wp.bool)
+        raise TypeError(f"Unsupported mask type: {type(mask)}")
+
+    # --- Fast path: all elements ---
+    if ids is None or (isinstance(ids, slice) and ids == slice(None)):
```
[the remark this answers is not in the document]
Comment (hujc7) [reply source/isaaclab/isaaclab/utils/warp/utils.py]: Fixed — removed the entire function (dead code, no callers on this branch).

File: source/isaaclab/isaaclab/utils/warp/utils.py [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+            # Use pre-allocated buffer - zero it and fill with indexed values
+            complete = wp.to_torch(out)
+            complete.zero_()
+            complete[ids] = value
+            return out
+        else:
+            # Create a complete data buffer from scratch
+            complete = torch.zeros((N, *value.shape[1:]), dtype=torch.float32, device=device)
+            complete[ids] = value
+            return wp.from_torch(complete, dtype=dtype)
```
[the remark this answers is not in the document]
Comment (hujc7) [reply source/isaaclab/isaaclab/utils/warp/utils.py]: Fixed — removed the entire function (dead code, no callers on this branch).

File: source/isaaclab_rl/isaaclab_rl/rsl_rl/vecenv_wrapper.py [hunk 3]
```diff
@@ -17,6 +18,9 @@
         ManagerBasedRLEnv,
     )
 
+    with contextlib.suppress(ImportError):
+        from isaaclab_experimental.envs import DirectRLEnvWarp, ManagerBasedRLEnvWarp
```
[the remark this answers is not in the document]
Comment (hujc7) [reply source/isaaclab_rl/isaaclab_rl/rsl_rl/vecenv_wrapper.py:22]: Low priority — not fixing in this PR. The contextlib.suppress pattern is consistent with the other RL wrappers (rl_games, sb3, skrl).

**Comment by hujc7** [comment]
@greptileai Review

File: source/isaaclab_experimental/isaaclab_experimental/envs/utils/spaces.py
Comment (AntoineRichard) [inline source/isaaclab_experimental/isaaclab_experimental/envs/utils/spaces.py:1]: Remove? :D

File: source/isaaclab_experimental/isaaclab_experimental/envs/utils/io_descriptors.py
Comment (AntoineRichard) [inline source/isaaclab_experimental/isaaclab_experimental/envs/utils/io_descriptors.py:1]: Remove? :D

File: source/isaaclab/isaaclab/utils/wrench_composer.py
Comment (AntoineRichard) [inline source/isaaclab/isaaclab/utils/wrench_composer.py:1]: This is still outputing forces in the body frame? If yes all good :)

**Review by AntoineRichard — COMMENTED** [review]
It looks mostly safe to merge. Thanks a lot @hujc7 !

File: source/isaaclab/isaaclab/utils/wrench_composer.py [hunk 4]
```diff
@@ -45,9 +45,18 @@ def __init__(self, asset: BaseArticulation | BaseRigidObject | BaseRigidObjectCo
...
+        # We intentionally avoid caching body_com_pose_w (a Tier 2 derived property) because
+        # it is lazily computed via a Python timestamp guard.  Saving the .data pointer at init
+        # time would freeze it at the initial value — subsequent steps would read stale COM
+        # world poses since nothing triggers the lazy recomputation.  Instead, we keep the two
+        # Tier 1 inputs (body_link_pose_w and body_com_pos_b) and let the wrench kernels
+        # compute the COM pose inline.  This is both correct in eager mode and CUDA-graph-
+        # capture safe (Tier 1 buffers are stable sim-bind pointers updated by the solver).
+        data = self._asset.data
+        if hasattr(data, "body_link_pose_w") and hasattr(data, "body_com_pos_b"):
+            self._body_link_pose_w = data.body_link_pose_w
```
[the remark this answers is not in the document]
Comment (hujc7) [reply source/isaaclab/isaaclab/utils/wrench_composer.py:58]: Intentional — both buffers needed for correct COM pose. Old code cached stale Tier 2 property. All IsaacLab asset types expose both.

File: source/isaaclab_experimental/isaaclab_experimental/envs/utils/spaces.py
Comment (hujc7) [reply source/isaaclab_experimental/isaaclab_experimental/envs/utils/spaces.py:1]: Already removed in PR #4905 — imports from stable now.

File: source/isaaclab_experimental/isaaclab_experimental/envs/utils/io_descriptors.py
Comment (hujc7) [reply source/isaaclab_experimental/isaaclab_experimental/envs/utils/io_descriptors.py:1]: Not a duplicate — warp-specific decorator for `func(env, out) -> None` obs terms. Stable expects `func(env) -> Tensor`. Descriptor dataclass reused from stable.

File: source/isaaclab/isaaclab/utils/wrench_composer.py
Comment (hujc7) [reply source/isaaclab/isaaclab/utils/wrench_composer.py:1]: Yes, body frame. The only change is computing COM pose inline from Tier 1 buffers (body_link_pose_w + body_com_pos_b) instead of caching the Tier 2 body_com_pose_w property. No change to output frame or kernel logic.

File: source/isaaclab/isaaclab/utils/warp/kernels.py [hunk 5]
```diff
@@ -459,45 +472,42 @@ def set_forces_and_torques_at_position_index(
...
         )
     # set the forces to the composed force, if the positions are provided, adds a torque to the composed torque
     # from the force at that position.
     if forces:
         # set the forces to the composed force
-        composed_forces_b[env_ids[tid_env], body_ids[tid_body]] = cast_force_to_link_frame(
-            forces[tid_env, tid_body],
-            wp.transform_get_rotation(link_poses[env_ids[tid_env], body_ids[tid_body]]),
-            is_global,
+        composed_forces_b[env_ids[tid_env], body_ids[tid_body]] = cast_force_to_com_frame(
```
[the remark this answers is not in the document]
Comment (hujc7) [reply source/isaaclab/isaaclab/utils/warp/kernels.py:503]: Pre-existing behavior — not introduced by this PR. Added a NOTE comment to both `set_*` variants documenting the overwrite caveat. No caller currently passes both torques and forces+positions.

**Comment by hujc7** [comment]
@greptileai Review

File: source/isaaclab_experimental/isaaclab_experimental/envs/mdp/events.py [hunk 6]
```diff
@@ -0,0 +1,614 @@
...
+        push_by_setting_velocity._scratch_vel = wp.zeros((env.num_envs,), dtype=wp.spatial_vectorf, device=env.device)
+        r = [velocity_range.get(key, (0.0, 0.0)) for key in ["x", "y", "z", "roll", "pitch", "yaw"]]
+        push_by_setting_velocity._lin_lo = wp.vec3f(r[0][0], r[1][0], r[2][0])
+        push_by_setting_velocity._lin_hi = wp.vec3f(r[0][1], r[1][1], r[2][1])
+        push_by_setting_velocity._ang_lo = wp.vec3f(r[3][0], r[4][0], r[5][0])
+        push_by_setting_velocity._ang_hi = wp.vec3f(r[3][1], r[4][1], r[5][1])
+
+    wp.launch(
+        kernel=_push_by_setting_velocity_kernel,
+        dim=env.num_envs,
```
[the remark this answers is not in the document]
Comment (hujc7) [reply source/isaaclab_experimental/isaaclab_experimental/envs/mdp/events.py]: Stale review — scratch buffer pattern was removed in latest push.

File: source/isaaclab_experimental/isaaclab_experimental/utils/warp/utils.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+
+@wp.func
+def wrap_to_pi(angle: float) -> float:
+    """Wrap input angle (in radians) to the range [-pi, pi)."""
+    two_pi = 2.0 * wp.pi
+    wrapped_angle = angle + wp.pi
+    # NOTE: Use floor-based remainder semantics to match torch's `%` for negative inputs.
+    wrapped_angle = wrapped_angle - wp.floor(wrapped_angle / two_pi) * two_pi
+    return wp.where((wrapped_angle == 0) and (angle > 0), wp.pi, wrapped_angle - wp.pi)
```
[the remark this answers is not in the document]
Comment (hujc7) [reply source/isaaclab_experimental/isaaclab_experimental/utils/warp/utils.py:138]: Docstring was incorrect. Fixed to [-pi, pi] to match stable isaaclab.utils.math.wrap_to_pi convention. Fix is in PR #4829 where the function is introduced.

File: source/isaaclab_experimental/isaaclab_experimental/envs/mdp/terminations.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+            out[i] = True
+
+
+def joint_pos_out_of_manual_limit(
+    env: ManagerBasedRLEnv, out, bounds: tuple[float, float], asset_cfg: SceneEntityCfg
+) -> None:
+    """Terminate when joint positions are outside configured bounds. Writes into ``out``."""
+    asset: Articulation = env.scene[asset_cfg.name]
+    assert asset_cfg.joint_mask is not None
+    assert asset.data.joint_pos.shape[1] == asset_cfg.joint_mask.shape[0]
```
[the remark this answers is not in the document]
Comment (hujc7) [reply source/isaaclab_experimental/isaaclab_experimental/envs/mdp/terminations.py]: Replaced with explicit raise ValueError in PR #4829 where these asserts are introduced.

**Comment by hujc7** [comment]
@greptileai Review

File: source/isaaclab_experimental/isaaclab_experimental/envs/mdp/terminations.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    lower: float,
+    upper: float,
+    out: wp.array(dtype=wp.bool),
+):
+    """2D kernel (num_envs, num_joints). ``out`` is pre-zeroed; only writes True."""
+    i, j = wp.tid()
+    if joint_mask[j]:
+        v = joint_pos[i, j]
+        if v < lower or v > upper:
+            out[i] = True
```
[the remark this answers is not in the document]
Comment (hujc7) [reply source/isaaclab_experimental/isaaclab_experimental/envs/mdp/terminations.py:99]: Acknowledged. All threads write the same value (`True`) to the same address — CUDA hardware serializes same-address stores within a warp and in practice this is safe even across warps for idempotent writes. Will restructure to a 1D kernel with inner loop in a follow-up for spec correctness.

**Comment by hujc7** [comment]
@greptileai Review

File: source/isaaclab_experimental/isaaclab_experimental/envs/mdp/events.py [hunk 10]
```diff
@@ -0,0 +1,628 @@
...
+        joint_pos[env_id, joint_id] = pos
+        joint_vel[env_id, joint_id] = vel
+
+    rng_state[env_id] = state
+
+
+# ---------------------------------------------------------------------------
+# Reset joints by offset
+# ---------------------------------------------------------------------------
+
```
[the remark this answers is not in the document]
Comment (hujc7) [reply source/isaaclab_experimental/isaaclab_experimental/envs/mdp/events.py]: Done. Removed the dead first definition.

**Comment by hujc7** [comment]
@greptileai Review

**Comment by hujc7** [comment]
Hi, @ooctipus. This PR still needs some attention to get merged.

**Comment by hujc7** [comment]
@greptileai Review

**Review by AntoineRichard — APPROVED** [review]

**Comment by hujc7** [comment]
## Post-merge validation — smoke + parity vs torch-MDP (Newton on both)

Ran two validation passes after merging the latest `origin/develop` (merge commit `0bb4fc525ad`, which pulled in #5265's dual-buffer WrenchComposer). Both on L40.

### 1. Smoke — 17/17 PASS (~11 min)

All 17 train envs registered by this PR construct, reset, step, and complete 2 training iters under `wp.ScopedCapture` (Newton's default `use_cuda_graph=True`). Zero `@capture_unsafe` violations; zero tracebacks.

```
python scripts/reinforcement_learning/rsl_rl/train.py \
  --task <env> --num_envs 16 --headless --max_iterations 2 --seed 42
```

### 2. Torch-MDP ↔ Warp-MDP parity — 17/17 PASS (~45 min)

Pair-parallel per env under NVIDIA MPS, 4096 envs, seed 42, max 300 iters. **Both sides forced onto Newton physics** (torch side via `presets=newton`) so the only variable is the MDP frontend. Early-PASS at iter 100/150 if reward ratio in [0.5, 2.0]; otherwise run to iter 300.

| Env | ratio (w/t) | torch rew / ep | warp rew / ep | decided @ iter |
|---|---|---|---|---|
| Cartpole | 1.00 | 4.94 / 300 | 4.96 / 300 | 100 |
| Cartpole-Direct | 1.00 | 294 / 299 | 295 / 299 | 100 |
| Ant | 1.22 | 60.6 / 931 | 73.8 / 945 | 100 |
| Ant-Direct | 1.49 | 5412 / 814 | 8039 / 878 | 150 |
| Humanoid | 1.62 | 82.0 / 752 | 132 / 871 | 150 |
| Humanoid-Direct | 1.40 | 806 / 153 | 1132 / 188 | 100 |
| Reach-Franka | \|2.0\| | -0.18 / 360 | 0.36 / 360 | 100 |
| Repose-Cube-Allegro-Direct | 0.93 | 103 / 272 | 95.4 / 257 | 150 |
| Velocity-Flat-Unitree-A1 | 1.47 | 22.2 / 964 | 32.7 / 975 | 150 |
| Velocity-Flat-Unitree-Go1 | 1.51 | 11.7 / 895 | 17.7 / 997 | 100 |
| Velocity-Flat-Unitree-Go2 | 1.86 | 9.76 / 994 | 18.2 / 952 | 100 |
| Velocity-Flat-Anymal-B | 1.33 | 14.3 / 990 | 19.0 / 958 | 100 |
| Velocity-Flat-Anymal-C | 1.65 | 10.4 / 935 | 17.1 / 964 | 100 |
| Velocity-Flat-Anymal-D | 1.34 | 5.93 / 846 | 7.97 / 971 | 100 |
| Velocity-Flat-Cassie | 1.66 | 23.6 / 951 | 39.2 / 961 | 150 |
| Velocity-Flat-G1 | 1.24 | -5.18 / 93 | -6.43 / 934 | 100 |
| Velocity-Flat-H1 | 1.13 | 29.6 / 980 | 33.6 / 1000 | 299 |

### Observations

- **Warp MDP consistently leads torch MDP at iter 100** (ratios 1.2–1.9 in most envs). Torch converges by iter 150–300 in every pair. Not a regression — consistent with Warp's MDP computation scaling better at 4096 envs.
- **G1 eplen asymmetry at iter 100**: torch=93 vs warp=934. Reward ratio 1.24 still passes the band (both negative, small-magnitude), but the episode-length gap is worth flagging if exact G1 parity is desired.
- **Antoine's dual-buffer WrenchComposer (#5265)** does not break any env and does not fire the `@capture_unsafe` guard during capture. Safe in the current call path because `compose_to_body_frame()` runs during `scene.write_data_to_sim()`, outside `NewtonManager._simulate()`'s ScopedCapture.

### Setup notes for reproducibility

- `isaaclab_ovphysx` must be installed (`python -m pip install -e source/isaaclab_ovphysx --no-deps`) — required by some torch direct envs.
- Warp envs (`Isaac-X-Warp-v0`) reject `presets=newton` (they are Newton-only and have no preset selector); only pass `presets=newton` to the torch side.
- MPS daemon enabled via `nvidia-cuda-mps-control -d` for pair-parallel runs.
